zetaclient/utils.go

```diff
@@ -76,3 +91,99 @@ func (t *DynamicTicker) UpdateInterval(newInterval uint64, logger zerolog.Logger
+		hex.EncodeToString(event.Message),
+		event.Raw.TxHash.Hex(),
+		event.Raw.BlockNumber,
+		1_500_000,
+		common.CoinType_ERC20,
+		event.Asset.String(),
+		ob.zetaClient.keys.GetOperatorAddress().String(),
+	), nil
+}
+
+func (ob *EVMChainClient) GetInboundVoteMsgForZetaSentEvent(event *zetaconnector.ZetaConnectorNonEthZetaSent) (types.MsgVoteOnObservedInboundTx, error) {
+	ob.logger.ExternalChainWatcher.Info().Msgf("TxBlockNumber %d Transaction Hash: %s Message : %s", event.Raw.BlockNumber, event.Raw.TxHash, event.Message)
+	destChain := common.GetChainFromChainID(event.DestinationChainId.Int64())
+	if destChain == nil {
+		ob.logger.ExternalChainWatcher.Warn().Msgf("chain id not supported  %d", event.DestinationChainId.Int64())
+		return types.MsgVoteOnObservedInboundTx{}, fmt.Errorf("chain id not supported  %d", event.DestinationChainId.Int64())
+	}
+	destAddr := clienttypes.BytesToEthHex(event.DestinationAddress)
+	if *destChain != common.ZetaChain() {
+		cfgDest, found := ob.cfg.GetEVMConfig(destChain.ChainId)
+		if !found {
+			return types.MsgVoteOnObservedInboundTx{}, fmt.Errorf("chain id not present in EVMChainConfigs  %d", event.DestinationChainId.Int64())
+		}
+		if strings.EqualFold(destAddr, cfgDest.ZetaTokenContractAddress) {
+			ob.logger.ExternalChainWatcher.Warn().Msgf("potential attack attempt: %s destination address is ZETA token contract address %s", destChain, destAddr)
+			return types.MsgVoteOnObservedInboundTx{}, fmt.Errorf("potential attack attempt: %s destination address is ZETA token contract address %s", destChain, destAddr)
+		}
+	}
+	return *GetInBoundVoteMessage(
+		event.ZetaTxSenderAddress.Hex(),
+		ob.chain.ChainId,
+		event.SourceTxOriginAddress.Hex(),
+		clienttypes.BytesToEthHex(event.DestinationAddress),
+		destChain.ChainId,
+		sdkmath.NewUintFromBigInt(event.ZetaValueAndGas),
+		base64.StdEncoding.EncodeToString(event.Message),
+		event.Raw.TxHash.Hex(),
+		event.Raw.BlockNumber,
+		event.DestinationGasLimit.Uint64(),
+		common.CoinType_Zeta,
+		"",
+		ob.zetaClient.keys.GetOperatorAddress().String(),
+	), nil
+}
+
+func (ob *EVMChainClient) GetInboundVoteMsgForTokenSentToTSS(txhash ethcommon.Hash, value *big.Int, receipt *ethtypes.Receipt, from ethcommon.Address, data []byte) *types.MsgVoteOnObservedInboundTx {
+	ob.logger.ExternalChainWatcher.Info().Msgf("TSS inTx detected: %s, blocknum %d", txhash.Hex(), receipt.BlockNumber)
+	ob.logger.ExternalChainWatcher.Info().Msgf("TSS inTx value: %s", value.String())
+	ob.logger.ExternalChainWatcher.Info().Msgf("TSS inTx from: %s", from.Hex())
+	message := ""
+	if len(data) != 0 {
+		message = hex.EncodeToString(data)
+	}
+	return GetInBoundVoteMessage(
+		from.Hex(),
+		ob.chain.ChainId,
+		from.Hex(),
+		from.Hex(),
+		common.ZetaChain().ChainId,
+		sdkmath.NewUintFromBigInt(value),
+		message,
+		txhash.Hex(),
+		receipt.BlockNumber.Uint64(),
+		90_000,
+		common.CoinType_Gas,
+		"",
+		ob.zetaClient.keys.GetOperatorAddress().String(),
+	)
+}
```

Comment:
I think we should unit tests (or create an issue to track writting unit tests) these methods since I may think a broken logic in the event parsing for inbound might ends up in incorrect interpretation of the data

Reply:
That makes sense, ill add an issue. 
However, the parsing logic is not new, the pr reuses existing parsing logic.